### ISSUE TITLE
Pyinstaller build for Linux (other OSes should follow after the ports are done)

### ISSUE THREAD

**Issue by MerlijnWajer** [JoeLametta/whipper#85]
We could make the distribution and usage of whipper easier by using Pyinstaller. It can take a python project, with lots of dlls and dependencies, and turn it into a single self-extracting binary. That may make distribution easier, and we could even ship our own cdparanoia/cdrdao (if we ever want/need to).

I've done something like this before, and it should in fact not be super hard.

Other platforms besides Linux will likely benefit more, but let's start with Linux.

**Comment by pierrrrrrre** [comment]
Does it work with C source code that we have to compile? (I'm thinking about the `accuraterip-checksum.c` program)

I was thinking about packaging and thought we could also give [snapcraft](http://snapcraft.io/) a try maybe? (it's Linux only and I don't think there are plans to port this to other OSes though)

**Comment by MerlijnWajer** [comment]
Yes. pyinstaller will include any amount of files or binaries, and possibly add them to your $PATH, etc. I have this working in a project, with morituri already.

**Comment by MerlijnWajer** [comment]
As for general packaging, I would go for the distro-native ways. E.g. on Gentoo an ebuild, the deb stuff for Debian, and so on. If this turns out to be annoying, and snapcraft solves those annoyances, fine.

My goal with pyinstaller is to have a simple downloadable way to get the package, also for Windows/OS-X (when we get there). I am sure distros do not want pyinstaller-like large files, as they will do the dependency management themselves.

**Comment by JoeLametta** [comment]
> As for general packaging, I would go for the distro-native ways. E.g. on Gentoo an ebuild, the deb stuff for Debian, and so on.

That's my opinion too.

**Comment by IvanDSM** [comment]
Agree with keeping it native, snappy/appimage/flatpak and the like can be a hassle sometimes since their package updates aren't synchronized with the distro's package updates, and therefore might leave an app outdated.

**Comment by pierrrrrrre** [comment]
snap is the way to go for Ubuntu.

> their package updates aren't synchronized with the distro's package updates, and therefore might leave an app outdated.

For snap, the main purpose is so that application developers can package and release their apps as needed and don't have to wait for approval by the distro staff. I think it's valuable in the case of a user-oriented software like whipper since by using snap you can release new versions quickly if needed, instead of having a very old version that has been validated once a long time ago and is completely outdated.

**Comment by RecursiveForest** [comment]
>That said, any packaging method will be better than the current install process :)

Were you not able to install it via your package manager? I don't think the current process is that arcane anymore, especially now that submodules are gone.

**Comment by pierrrrrrre** [comment]
I installed it a month ago following the README. It was more or less straightforward, but it's definitely not user-friendly. What I meant with my previous comment is that if it's in a PPA for instance, Ubuntu users can install it very easily (and if it's a snap, it can actually be pushed to the software center, making it even easier to install).

**Comment by MerlijnWajer** [comment]
I understand. Do you understand what pyinstaller does? It creates a single binary - with *everything* in it. It's a simple download and run. No dependencies to install, no nothing.

Distros can still package it in any way they want, include 'snaps', but let's make separate issues for that.

**Comment by pierrrrrrre** [comment]
> I understand. Do you understand what pyinstaller does? It creates a single binary - with everything in it. It's a simple download and run. No dependencies to install, no nothing.

Sounds good to me!

**Comment by gthieleb** [comment]
First steps that should be done is making whipper installable via pypi and therefore available within python virtualenvs without hassle. The step making distro related packages is then an easy step afterwards, as some distros have helper tools e.g. debian's [dh_pythonX](https://wiki.debian.org/Python/FAQ#What_are_dh_python2_and_dh_python3.3F) helper tools or pythons builtin [bdist_rpm](https://docs.python.org/2.0/dist/creating-rpms.html) flag. Pypi support is also a todo step for dependent packages like the currently broken setup.py of PyCMDB package and the missing requirements.txt file in whipper itself!

With pip/virtualenv support a much broader range of developers could be reached!
